src/ImmutableProperties.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Codelicia\Immutable;
+
+use ReflectionObject;
+use ReflectionProperty;
+use ReflectionType;
+use RuntimeException;
+use function array_key_exists;
+use function class_implements;
+use function get_class;
+use function gettype;
+use function in_array;
+use function interface_exists;
+use function is_object;
+use function spl_object_hash;
+use function sprintf;
+
+trait ImmutableProperties
+{
+    /** @return string[][]|ReflectionProperty[][] */
+    protected function &reflectionProperty(): array
+    {
+        static $reflectionProperty = [];
+
+        if (! array_key_exists(spl_object_hash($this), $reflectionProperty)) {
```

Comment:
The properties stay the same based on object type, so we don't need them per-instance. I suggest replacing `spl_object_hash()` with `get_class($this)` instead

Reply:
But then we cannot have more then one immutable object. I wouldn’t be able to have a `User(name=John)` and a `User(name=Josh)` at the same time as both will have the same values.